tombstones.go

```diff
@@ -150,36 +150,46 @@ func readTombstones(dir string) (memTombstones, error) {
 			return nil, d.err()
 		}
 
-		stonesMap.add(k, Interval{mint, maxt})
+		stonesMap.addInterval(k, Interval{mint, maxt})
 	}
 
 	return stonesMap, nil
 }
 
-type memTombstones map[uint64]Intervals
-
-var emptyTombstoneReader = memTombstones{}
+type memTombstones struct {
+	mts map[uint64]Intervals
+	mtx sync.RWMutex
+}
 
-// EmptyTombstoneReader returns a TombstoneReader that is always empty.
-func EmptyTombstoneReader() TombstoneReader {
-	return emptyTombstoneReader
+func NewMemTombstones() *memTombstones {
+	return &memTombstones{mts: make(map[uint64]Intervals)}
 }
 
-func (t memTombstones) Get(ref uint64) (Intervals, error) {
-	return t[ref], nil
+func (t *memTombstones) Get(ref uint64) (Intervals, error) {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.mts[ref], nil
 }
 
-func (t memTombstones) Iter(f func(uint64, Intervals) error) error {
-	for ref, ivs := range t {
+func (t *memTombstones) Iter(f func(uint64, Intervals) error) error {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	for ref, ivs := range t.mts {
 		if err := f(ref, ivs); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-func (t memTombstones) add(ref uint64, itv Interval) {
-	t[ref] = t[ref].add(itv)
+// addInterval to an existing memTombstones
+func (t *memTombstones) addInterval(ref uint64, itvs ...Interval) *memTombstones {
```

Comment:
Why are we returning ourselves again? Is this only so the changed test can build input test cases more easily?
In general we are not returning the object of a method again unless its a copy and the original object wasn't mutated.

I think for the test case is fine to keep setting the map directly since we are just initializing it.

Reply:
@fabxc  Yes, if there is no other cases, I will make it not returning itself.
Do you mean we construct the object directly?
```
		tombstones: &memTombstones{mts: map[uint64]Intervals{
			1: Intervals{{1, 3}},
			2: Intervals{{1, 3}, {6, 10}},
			3: Intervals{{6, 10}},
		}},
```

Reply:
Yes, that would be more consistent with what we have elsewhere.